api/cluster.go

```diff
@@ -312,103 +312,136 @@ func (s *Server) peerQuery(ctx context.Context, data cluster.Traceable, name, pa
 // name:         name to be used in logging & tracing
 // path:         path to request on
 func (s *Server) peerQuerySpeculative(ctx context.Context, data cluster.Traceable, name, path string) (map[string]PeerResponse, error) {
-	peerGroups, err := cluster.MembersForSpeculativeQuery()
-	if err != nil {
-		log.Error(3, "HTTP peerQuery unable to get peers, %s", err)
-		return nil, err
-	}
-	log.Debug("HTTP %s across %d instances", name, len(peerGroups)-1)
+	result := make(map[string]PeerResponse)
 
-	reqCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	responseChan, errorChan := s.peerQuerySpeculativeChan(ctx, data, name, path)
 
-	originalPeers := make(map[string]struct{}, len(peerGroups))
-	receivedResponses := make(map[int32]struct{}, len(peerGroups))
+	for resp := range responseChan {
+		result[resp.peer.GetName()] = resp
+	}
 
-	responses := make(chan struct {
-		shardGroup int32
-		data       PeerResponse
-		err        error
-	}, 1)
+	err := <-errorChan
+	return result, err
+}
 
-	askPeer := func(shardGroup int32, peer cluster.Node) {
-		log.Debug("HTTP Render querying %s%s", peer.GetName(), path)
-		buf, err := peer.Post(reqCtx, name, path, data)
+// peerQuerySpeculativeChan takes a request and the path to request it on, then fans it out
+// across the cluster. If any peer fails requests to other peers are aborted. If enough
+// peers have been heard from (based on speculation-threshold configuration), and we
+// are missing the others, try to speculatively query other members of the shard group.
+// ctx:          request context
+// data:         request to be submitted
+// name:         name to be used in logging & tracing
+// path:         path to request on
+// resultChan:   channel to put responses on as they come in
+func (s *Server) peerQuerySpeculativeChan(ctx context.Context, data cluster.Traceable, name, path string) (<-chan PeerResponse, <-chan error) {
+	resultChan := make(chan PeerResponse)
+	errorChan := make(chan error, 1)
 
-		select {
-		case <-ctx.Done():
-			return
-		default:
-			// Not canceled, continue
-		}
+	go func() {
+		defer close(errorChan)
+		defer close(resultChan)
```

Comment:
minor thought: would it be cleaner/simpler to just have 1 return channel? we could return PeerResponse along with its error

Reply:
Possibly. I like the 2 channels, because you can just loop over responses until it is closed, then quickly try to read into an err var for the return value. See [this usage](https://github.com/grafana/metrictank/pull/1021/files#diff-841baed3f3a566ff48a5a4a427defb30R920)

Reply:
fair enough